### PR TITLE
Use a manifest file so as to not have results in memory during benchmarking.

### PR DESCRIPTION
**Pull request by vext01**
NOT EVEN CLOSE TO BEING READY :)

I suppose cookie support looks a bit like this. Please comment on the rough approach.

I've realised that anyone wishing to use --resume manually, will have to make sure the right cookie is in place.

I also need to try to detect a stale cookie. I guess it could happen.

### CODE REVIEW

File: krun/scheduler.py [hunk 1]
```diff
@@ -149,6 +148,10 @@ def jobs_until_eta_known(self):
     def add_job(self, job):
         self.work_deque.append(job)
 
+    def clear_jobs(self):
+        del self.work_deque
+        self.work_deque = deque
```
Comment (ltratt) [inline krun/scheduler.py]: Reference of an undefined variable `deque`?

**Comment by ltratt** [comment]
Apart from that one comment, the logic makes sense to me.

In terms of resuming, I think we just need to say "if you interrupt Krun, you need to keep the status cookie unchanged. Removing or editing it leads to undefined behaviour."

**Comment by snim2** [comment]
This seems sensible to me. One possible corner case: if something goes wrong with writing the status cookie, I think Krun will think it is perpetually on the first task in the schedule and go into an infinite loop.

**Comment by ltratt** [comment]
If the write fails, I think it'll raise an exception, and then stop doing anything?

**Comment by snim2** [comment]
I mean if the read fails...

File: krun/scheduler.py (hunk 1 above)
Comment (vext01) [inline krun/scheduler.py]: Oops. It should be `deque()`

**Comment by ltratt** [comment]
I don't see how we can do anything if the cookie keeps disappearing (which means the user is either destroying it, or the hard disk is broken). IMHO, that comes strictly under the "don't do that" heading.

**Comment by vext01** [comment]
I'm happy to say "leave the cookie alone" if you want resume. But I wonder if we should kill the results stripper functionality, which clearly isn't going to be easy to use anymore.

I'm starting to think that killing little-used code in Krun can only be a good thing at this point.

**Comment by ltratt** [comment]
@vext01 I'm all for deleting code, but it might be worth thinking how much testing the deletion will require.

**Comment by vext01** [comment]
@snim2 Any thoughts?

**Comment by snim2** [comment]
I think if there is no cookie, carry on, if the read fails for some other reason we should abort.

**Comment by vext01** [comment]
About deleting the result stripper I mean.

**Comment by snim2** [comment]
Well, it was you that wanted that functionality :)

I think it is OK to delete it if you feel we have good test coverage.

**Comment by vext01** [comment]
Ironically, I've never used that functionality.

**Comment by vext01** [comment]
I wonder if there is any merit in having the cookie contain one line for each proc. exec, and then once the proc.exec is done, the line is marked as done (or removed). This means we don't have to repeatedly regenerate the schedule after each run, which means the scary code in `build_shcedule` can go.

Thoughts? @snim2

**Comment by vext01** [comment]
Correct, but we never need to decide what has already been run.

**Comment by vext01** [comment]
Suggest I try it out and see how we go?

**Comment by snim2** [comment]
Makes sense

**Comment by vext01** [comment]
It touches the ETA code, but I hope it isn't hard to fix. (famous last words?)

**Comment by vext01** [comment]
Still not ready, but here is most of the code (doesn't yet work).

On first run, Krun writes a manifest file to disk that looks like this:
```
eta_avail_idx=-1
keys
O dummy:C:default-c
O nbody:C:default-c
O dummy:CPython:default-python
O nbody:CPython:default-python
O dummy:C:default-c
O nbody:C:default-c
O dummy:CPython:default-python
O nbody:CPython:default-python
```

`keys` signifies the beginning of jobs, then each line is a `flag key` pair. Flags let us know the status of the benchmarking session:

 * O - outstanding
 * C - complete
 * E - errored
 * S - skip

The idea was to never such the whole file into memory, but to simply update the flags as required.

This means:

 * We only build the schedule once, and some of the scary logic disappears.
 * The only time you should have no manifest file is when you run for the first time, so it is easy to detect when something went wrong.

Downsides:
 * The change *is* invasive.

[The reason this doesn't work ATM is OpenBSD does not support writing to an arbitrary offset in an existing file (AFAICS). `a` mode for open always appends to the end of the file. This means I can't update the flags in the manifest in the way I intended. So we may end up having to suck the file into memory after all, boo.]

In an ideal world this change is nice, and clearly superior, but I fear it is too invasive for us now. What do you think people? Maybe after all I should just revert to the old way where we build a schedule after each proc. exec.

**Comment by snim2** [comment]
3 files (minus tests) seems reasonable in terms of invasiveness, although this is the tricky bit of the code. The manifest itself looks good to me.

**Comment by vext01** [comment]
So I think you are telling me that this is worthwhile to continue working on.

**Comment by snim2** [comment]
I am!

**Comment by ltratt** [comment]
Fine with me.

I think you should be able to seek to a position and then write a limited number of bytes without issue?

**Comment by vext01** [comment]
I thought I could too. Will investigate...

**Comment by vext01** [comment]
```
vext01@bencher5:/tmp$ cat a.py 
import os

with open("testfile", "w") as fh:
    fh.write("123456")
os.system("hexdump -C  testfile")

with open("testfile", "a+") as fh:
    fh.seek(3)
    fh.write("X")
os.system("hexdump -C  testfile")
vext01@bencher5:/tmp$ python2.7 a.py 
00000000  31 32 33 34 35 36                                 |123456|
00000006
00000000  31 32 33 34 35 36 58                              |123456X|
00000007
```

**Comment by vext01** [comment]
This works:

```
import os

with open("testfile", "w") as fh:
    fh.write("123456")
os.system("hexdump -C  testfile")

fd = os.open("testfile", os.O_WRONLY)
os.lseek(fd, 3, os.SEEK_SET)
os.write(fd, "X")
os.close(fd)
os.system("hexdump -C  testfile")
vext01@bencher5:/tmp$ python2.7 a.py 
00000000  31 32 33 34 35 36                                 |123456|
00000006
00000000  31 32 33 58 35 36                                 |123X56|
00000006
```

**Comment by ltratt** [comment]
Good ;) If it hadn't worked, I think none of the database out there would be working properly on OpenBSD!

**Comment by vext01** [comment]
yeah, so file operations are not flexible enough, we have to use fds.

**Comment by vext01** [comment]
[This is the same on linux even]

**Comment by snim2** [comment]
Fixes #253

**Comment by ltratt** [comment]
OK, I'll need to be the final approver for this one, as I'm the person who hasn't been involved ;) Let me know when it's ready. It might be worth coordinating some squishing between the two of you before that point, to make the reviewing a little easier, but that's up to you both.

**Comment by vext01** [comment]
Tests fixed -- @snim2 if you do your squashes now, I will review the remainder and decide how/if to squash.

**Comment by snim2** [comment]
OK @vext01 your turn!

**Comment by vext01** [comment]
I've squashed this down best I can.

You OK to start reviewing this @ltratt.

(Apologies this is so large)

File: README.md [hunk 2]
```diff
@@ -406,8 +368,7 @@ miscompiled, you may want to re-run all benchmarks for the troublesome VM only.
 Of course, under ideal circumstances, you would collect all results in one go.
 
 You can use the `--strip-results` mode to strip results from your result file.
```
Comment (vext01) [inline README.md]: `--strip results` was removed. We should kill all mentions of it.

File: etc/rc.local.openbsd [hunk 3]
```diff
@@ -26,5 +26,4 @@ PYTHON=/usr/local/bin/python2.7
 # Don't change anything below here
 
 env PATH=${PATH}:/usr/local/bin sudo -u ${INITIAL_USER} sh -c \
-    "cd ${EXPERIMENT_DIR} && ${PYTHON} ${KRUN_DIR}/krun.py \
-   --resume --reboot --started-by-init ${CONFIG_FILE}" &
+    "cd ${EXPERIMENT_DIR} && ${PYTHON} ${KRUN_DIR}/krun.py --hardware-reboots ${CONFIG_FILE}" &
```
Comment (vext01) [inline etc/rc.local.openbsd]: We need to update the example `rc.local` files, but shall we do that as a separate PR?

File: examples/example.krun [hunk 4]
```diff
@@ -45,7 +45,7 @@ VMS = {
 
 BENCHMARKS = {
     'dummy': 1000,
-    'nbody': 1000,
+    #'nbody': 1000,
```
Comment (ltratt) [inline examples/example.krun]: Let's delete this line.

File: krun.py [hunk 5]
```diff
@@ -192,11 +177,10 @@ def main(parser):
         util.print_session_info(config)
         return
 
-    if args.strip_results:
-        util.strip_results(config, args.strip_results)
-        return
+    on_first_invokation = not (os.path.isfile(ManifestManager.PATH) and
```
Comment (vext01) [inline krun.py]: Did we somwhere check that if a manifest file exists, so should a results file, and the other way around?

File: krun/scheduler.py [hunk 6]
```diff
@@ -37,6 +36,132 @@ class ScheduleEmpty(Exception):
...
+        offset = 0
+
+        # Parse manifest header
+        for line in fh:
+            offset += len(line)
+            line = line.strip()
+            if line == "keys":
+                break
+            else:
+                key, val = line.split("=")
```
Comment (ltratt) [inline krun/scheduler.py]: This would be clearer and shorter as:
```python
assert key == "eta_avail_idx"
self.eta_avail_idx = int(val)
```

File: krun.py [hunk 7]
```diff
@@ -275,8 +252,8 @@ def inner_main(mailer, config, args):
         platform.starting_temperatures = current.starting_temperatures
     else:
         # Touch the config file to update its mtime. This is required
-        # by resume-mode which uses the mtime to determine the name of
-        # the log file, should this benchmark be resumed.
+        # by when resuming a partially complete benchmark, in which case Krun
```
Comment (vext01) [inline krun.py]: A partially complete benchmark *session*, right?

File: krun/platform.py [hunk 8]
```diff
@@ -654,8 +654,16 @@ def __init__(self, mailer, config):
         self.temp_sensor_map = None
         UnixLikePlatform.__init__(self, mailer, config)
         self.num_cpus = self._get_num_cpus()
+        self.num_per_core_measurements = self._get_num_per_core_measurements()
+
         self.virt_what_cmd = None  # set later
 
+    def _get_num_per_core_measurements(self):
+        # For all systems apart from travis we expect per-core measurements
+        if os.environ.get("TRAVIS") == "true":
```
Comment (vext01) [inline krun/platform.py]: Note -- if this kind of thing happens much more (I hope not), we should make a `class TravisPlatform(DebianLinuxPlatform)`. Not now though.

File: krun/results.py [hunk 9]
```diff
@@ -93,10 +93,41 @@ def read_from_file(self, results_file):
...
+            # Check the length of the in-process interations in results
+            # XXX Check all core counts are the same
+            # XXX Check all exec counts are the same
+            expect_len = None
+            for exec_idx in xrange(len(execs)):
+                one_exec = execs[exec_idx]
+                if expect_len is None:
+                    expect_len = len(one_exec)
+
+                #if not all([len(x) == expect_len for x in self.core_cycle_counts[key][exec_idx]]):
```
Comment (ltratt) [inline krun/results.py]: Delete line.

File: krun/results.py [hunk 10]
```diff
@@ -93,10 +93,41 @@ def read_from_file(self, results_file):
...
+                for core in self.core_cycle_counts[key][exec_idx]:
+                    if len(core) != expect_len:
+                        fatal("inconsistent results: core_cycle_counts length")
+
```
Comment (ltratt) [inline krun/results.py]: Extra blank line.

File: krun/scheduler.py [hunk 11]
```diff
@@ -279,160 +314,178 @@ def _make_pre_post_cmd_env(self):
...
-
-            if self.reboot and len(self) > 0:
-                info("Reboot in preparation for next execution")
-                self._reboot()
+        if self.on_first_invokation:
+            info("Reboot prior to first execution")
+            self._reboot()
 
-        self.platform.save_power()
+        # If we get here, this is a real run, with a benchmark about to
```
Comment (vext01) [inline krun/scheduler.py]: I'm really happy with the simplification of `run()`, it's much easier to understand now.

File: krun/scheduler.py [hunk 12]
```diff
@@ -350,131 +346,134 @@ def run(self):
...
-            else:
-                assert False
+            info("Reboot in preparation for next execution")
+            self._reboot(self.manifest.total_num_execs)
 
-            if self.reboot and self.manifest.num_execs_left > 0:
-                info("Reboot in preparation for next execution")
-                self._reboot(self.manifest.total_num_execs)
+        elif self.manifest.num_execs_left < 0:
+            assert False
```
Comment (ltratt) [inline krun/scheduler.py]: This is is not a good `assert False` -- it should be somewhere else (e.g. before the `if`).

File: krun/scheduler.py [hunk 13]
```diff
@@ -350,131 +346,134 @@ def run(self):
...
-        self.platform.save_power()
+        elif self.manifest.num_execs_left ==  0:
+            self.platform.save_power()
 
-        info("Done: Results dumped to %s" % self.config.results_filename())
-        err_msg = "Errors/warnings occurred -- read the log!"
-        if self.results.error_flag:
-            warn(err_msg)
+            info("Done: Results dumped to %s" % self.config.results_filename())
+            err_msg = "Errors/warnings occurred -- read the log!"
```
Comment (ltratt) [inline krun/scheduler.py]: This doesn't make sense to me. a) if There are no execs why would we possibly raise a "error occured" message? b) If `error_flag` isn't set, why construct `err_msg`?

File: krun/tests/test_scheduler.py [hunk 14]
```diff
@@ -46,13 +43,12 @@ def dummy_reboot(self, _):
 
     reboots = 0
     while True:
-        resume = False
+        on_first_invokation = True
```
Comment (ltratt) [inline krun/tests/test_scheduler.py]: "invocation"

Comment (snim2) [inline krun/tests/test_scheduler.py]: oops

File: krun.py [hunk 15]
```diff
@@ -206,36 +190,29 @@ def main(parser):
     mailer = Mailer(mail_recipients, max_mails=config.MAX_MAILS)
 
     try:
-        inner_main(mailer, config, args)
+        inner_main(mailer, on_first_invokation, config, args)
     except util.FatalKrunError as e:
         subject = "Fatal Krun Exception"
         mailer.send(subject, e.args[0], bypass_limiter=True)
         util.run_shell_cmd_list(config.POST_EXECUTION_CMDS)
```
Comment (ltratt) [inline krun.py]: I think this should be:
```python
     except util.FatalKrunError as e:
          util.run_shell_cmd_list(config.POST_EXECUTION_CMDS)
          subject = "Fatal Krun Exception"
          mailer.send(subject, e.args[0], bypass_limiter=True)
```

Comment (ltratt) [inline krun.py]: I also wonder if `run_shell_cmd_list` should have an argument `try_all` (or similar) which, even if a command fails, tries to run the subsequent commands?

**Comment by ltratt** [comment]
OK, those are my initial comments.

File: examples/example.krun (hunk 4 above)
Comment (vext01) [inline examples/example.krun]: I'm confused -- this line is not commented on the head of the branch... GH bug?

File: etc/rc.local.openbsd (hunk 3 above)
Comment (vext01) [inline etc/rc.local.openbsd]: #264

File: krun/results.py (hunk 9 above)
Comment (vext01) [inline krun/results.py]: fixed in a later commit.

File: examples/example.krun (hunk 4 above)
Comment (vext01) [inline examples/example.krun]: fixed in a later commit.

File: krun/results.py (hunk 10 above)
Comment (vext01) [inline krun/results.py]: fixed in a later commit.

File: krun/scheduler.py (hunk 13 above)
Comment (vext01) [inline krun/scheduler.py]: If there was an error at any point in the session a benchmark has failed, a message is printed right at the end. This is mostly just useful for when you are testing, as in a real benchmarking session you would have received an email.

Comment (ltratt) [inline krun/scheduler.py]: OK, that addresses point a) (but not b)).

File: krun/tests/test_scheduler.py (hunk 14 above)
Comment (ltratt) [inline krun/tests/test_scheduler.py]: Das ist England ('16).

File: krun.py (hunk 15 above)
Comment (vext01) [inline krun.py]: Up until now, we have allowed commands to fail by having `| true` in the config. Is that enough?

Comment (vext01) [inline krun.py]: As for the re-order, let's do that in the pre/post hook fix branch.

Comment (ltratt) [inline krun.py]: Ah, yes, the `| true` hack is fine. Good point.

Comment (ltratt) [inline krun.py]: I don't think we should leave the reordering though -- it's a one line fix, and we'll forget it otherwise.

**Comment by vext01** [comment]
I think that covers it. One remaining comment outstanding for @snim2 regarding the consistency of the manifest and results file existing (or not).

**Comment by ltratt** [comment]
I'd like to reorder the `try/except` thing before we merge. Once that's done, I'm fine with this, and it can be squashed as needed.

[Note that tests are failing ATM though.]

**Comment by ltratt** [comment]
OK, once the tests have passed, please squash, then I can merge.

**Comment by vext01** [comment]
Will wait for Sarah before any squashing.

**Comment by snim2** [comment]
@vext01 I'm going to answer the question here. We have a number of checks in `krun.py` that deal with all of this, but they aren't all next to each other. Let me pull them out here so you can see:

```python
    on_first_invocation = not (os.path.isfile(ManifestManager.PATH) and
                               os.stat(ManifestManager.PATH).st_size > 0)

    ...

    out_file = config.results_filename()  # This is the results file
    out_file_exists = os.path.exists(out_file)

    if out_file_exists and not os.path.isfile(out_file):
        util.fatal(
            "Output file '%s' exists but is not a regular file" % out_file)

    if out_file_exists and on_first_invocation:
        util.fatal("Output results file '%s' already exists. "
                   "Move the file away before running Krun." % out_file)

    if not out_file_exists and not on_first_invocation:
        util.fatal("No results file to resume. Expected '%s'" % out_file)

    ...

    if not on_first_invocation:
        # output file must exist, due to check above
        assert(out_file_exists)
        current = Results(config, platform, results_file=out_file)
```

**Comment by vext01** [comment]
Alright -- will squash now. Thanks

**Comment by vext01** [comment]
I think that's the best squashing I can do without pain.

**Comment by ltratt** [comment]
We can definitely squash the "in an error case" commit away easily. The others are OK as-is though.

**Comment by vext01** [comment]
I kept it separate as it's not related to the manifest changes.

**Comment by ltratt** [comment]
Point taken.
